File: tweaks/tweaks/report/ac_principal_query_filters/ac_principal_query_filters.js
// Copyright (c) 2026, Erick W.R. and contributors
// For license information, please see license.txt

frappe.query_reports["AC Principal Query Filters"] = {
    filters: [],
    filter_data: {},
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        
        // Format the filter_name column to show link to Query Filter
        if (column.fieldname === "filter_name" && data.query_filter) {
            const link_url = frappe.utils.get_form_link("Query Filter", data.query_filter);
            return `<a href="${link_url}">${frappe.utils.escape_html(value)}</a>`;
        }
        
        // Format the user column to show full name with link to user profile
        if (column.fieldname === "user" && data.user_id) {
            const link_url = frappe.utils.get_form_link("User", data.user_id);
            return `<a href="${link_url}">${frappe.utils.escape_html(value)}</a>`;
        }
        
        // Format the reference_docname column to show link to the referenced document
        if (column.fieldname === "reference_docname" && data.reference_doctype && data.reference_docname) {
            const link_url = frappe.utils.get_form_link(data.reference_doctype, data.reference_docname);
            return `<a href="${link_url}">${frappe.utils.escape_html(value)}</a>`;
        }
        
        if (column.fieldname === "filters" && data.filters) {
            frappe.query_reports["AC Principal Query Filters"].filter_data[data.query_filter] = {
                filters: data.filters,
                filters_type: data.filters_type
            };
            const escaped_name = data.query_filter.replace(/'/g, "\\'");
            return `
            <button class="btn btn-default btn-xs" onclick="frappe.query_reports['AC Principal Query Filters'].show_filter_dialog('${escaped_name}')">
                View Filter
            </button>
            `;
        }
        
        return value;
    },
    "show_filter_dialog": function(name) {
        const filter_info = this.filter_data[name];
        
        if (!filter_info) {
            frappe.msgprint(__("Filter not found"));
            return;
        }
        
        // Determine the code editor language based on filter type
        let language = 'Python';
        if (filter_info.filters_type === 'SQL') {
            language = 'SQL';
        } else if (filter_info.filters_type === 'JSON') {
            language = 'JSON';
        }
        
        const dialog = new frappe.ui.Dialog({
            title: __('Filter: {0}', [name]),
            fields: [{
                fieldtype: 'Code',
                fieldname: 'filters',
                label: __('Filter ({0})', [filter_info.filters_type]),
                options: language,
                default: filter_info.filters,
            }],
            size: 'large'
        });
        
        dialog.show();
    }
};

File: tweaks/tweaks/report/ac_principal_query_filters/ac_principal_query_filters.py
# Copyright (c) 2026, Erick W.R. and contributors
# For license information, please see license.txt

import frappe
from frappe import _

from tweaks.tweaks.doctype.ac_rule.ac_rule_utils import get_principal_filter_sql


def execute(filters=None):
    """
    AC Principal Query Filters Report

    Shows each principal query filter (User, Role, User Group, Role Profile)
    with the SQL used to determine users and lists all matching users.

    Report Structure:
        - Rows: One row per (Query Filter, User) combination
        - Columns: Query Filter, Reference DocType, Reference DocName, Filter Type, Filter, SQL, User

    The report helps understand which users are matched by each principal filter.
    Includes Query Filter properties and a button to view filter code in a dialog.
    """
    columns = get_columns()
    data = get_data(filters)
    return columns, data


def get_columns():
    """Define report columns"""
    return [
        {
            "fieldname": "filter_name",
            "label": _("Filter Name"),
            "fieldtype": "Data",
            "width": 200,
        },
        {
            "fieldname": "query_filter",
            "label": _("Query Filter"),
            "fieldtype": "Link",
            "options": "Query Filter",
            "hidden": 1,
        },
        {
            "fieldname": "user",
            "label": _("User"),
            "fieldtype": "Data",
            "width": 200,
        },
        {
            "fieldname": "user_id",
            "label": _("User ID"),
            "fieldtype": "Link",
            "options": "User",
            "hidden": 1,
        },
        {
            "fieldname": "reference_doctype",
            "label": _("Reference DocType"),
            "fieldtype": "Link",
            "options": "DocType",
            "width": 150,
        },
        {
            "fieldname": "reference_docname",
            "label": _("Reference DocName"),
            "fieldtype": "Data",
            "width": 150,
        },
        {
            "fieldname": "filters_type",
            "label": _("Filter Type"),
            "fieldtype": "Data",
            "width": 100,
        },
        {
            "fieldname": "filters",
            "label": _("Filter"),
            "fieldtype": "Code",
            "width": 120,
        },
        {
            "fieldname": "sql",
            "label": _("SQL"),
            "fieldtype": "Data",
            "width": 300,
        },
    ]


def get_data(filters):
    """Get data for the report"""

    # Get all Query Filters of relevant types
    query_filters = frappe.get_all(
        "Query Filter",
        filters={
            "reference_doctype": ["in", ["User", "Role", "User Group", "Role Profile"]],
            "disabled": 0,
        },
        fields=[
            "name",
            "filter_name",
            "reference_doctype",
            "reference_docname",
            "filters_type",
            "filters",
        ],
        order_by="name",
    )

    data = []

    for qf_info in query_filters:
        # Load the full Query Filter document
        query_filter = frappe.get_doc("Query Filter", qf_info.name)

        # Get the SQL for this filter
        try:
            filter_sql = get_principal_filter_sql(query_filter)

            # Remove line breaks from SQL
            filter_sql_display = (
                filter_sql.replace("\n", " ").replace("\r", " ").strip()
            )

            # Get users matching this filter
            if filter_sql:
                users = frappe.db.sql(
                    f"""
                    SELECT DISTINCT `name`, `full_name`
                    FROM `tabUser`
                    WHERE {filter_sql} AND enabled = 1
                    ORDER BY `name`
                    """,
                    as_dict=1,
                )
            else:
                users = []

            # Create a row for each user
            for user in users:
                data.append(
                    {
                        "filter_name": qf_info.filter_name,
                        "query_filter": qf_info.name,
                        "reference_doctype": qf_info.reference_doctype,
                        "reference_docname": qf_info.reference_docname,
                        "filters_type": qf_info.filters_type,
                        "filters": qf_info.filters,
                        "sql": filter_sql_display,
                        "user": user["full_name"],
                        "user_id": user["name"],
                    }
                )

        except Exception as e:
            # If there's an error getting SQL or users, log it and skip this filter
            frappe.log_error(
                f"Error processing Query Filter {qf_info.name}: {str(e)}",
                "AC Principal Query Filters Report",
            )
            continue

    return data
